exe-common/src/sync.rs

```diff
@@ -659,3 +660,404 @@ pub fn get_chain_state_at_end_of(
+                    epoch: date.get_epochid(),
+                    chain_difficulty: ChainDifficulty::from(date.slot_number() as u64),
+                },
+                BlockHeaderAttributes(cbor_event::Value::U64(0)),
+            );
+            Block::BoundaryBlock(boundary::Block {
+                header: boundary_header,
+                body: boundary_body,
+                extra: cbor_event::Value::U64(0),
+            })
+        }
+
+        pub fn create_block(&mut self) {
+            let date = if self.tip == self.genesis_data.genesis_prev {
+                BlockDate::from(0 as EpochId)
+            } else {
+                let prev_date = self.blocks.get(&self.tip).unwrap().header().blockdate();
+                // BlockDate::next() does *not* handle wrap around, just a TODO there...
+                if prev_date.slotid().is_some()
+                    && prev_date.slotid().unwrap() >= SLOTS_PER_EPOCH - 1
+                {
+                    BlockDate::from(prev_date.get_epochid() + 1 as EpochId)
+                } else {
+                    prev_date.next()
+                }
+            };
+            let block = if date.is_boundary() {
+                self.make_boundary_block(&date)
+            } else {
+                let body = normal::Body::new(
+                    normal::TxPayload::empty(),
+                    normal::SscPayload::fake(),
+                    normal::DlgPayload(cbor_event::Value::U64(self.nonce)),
+                    update::UpdatePayload {
+                        proposal: None,
+                        votes: Vec::new(),
+                    },
+                );
+                let extra = cbor_event::Value::U64(0);
+                self.nonce += 1;
+
+                let body_proof = normal::BodyProof::generate_from_body(&body);
+                let slot_id = match date {
+                    BlockDate::Boundary(_) => unreachable!(),
+                    BlockDate::Normal(epochslot) => epochslot,
+                };
+                let chain_difficulty = ChainDifficulty::from(date.slot_number() as u64 + 1);
+                let header_extra_data = HeaderExtraData::new(
+                    BlockVersion::new(1, 0, 0),
+                    SoftwareVersion::new("rollbacktest", 0).unwrap(),
+                    BlockHeaderAttributes(cbor_event::Value::U64(0)),
+                    hash::Blake2b256::new(&cbor!(&extra).unwrap()),
+                );
+                let boot_stakeholder = self
+                    .genesis_data
+                    .boot_stakeholders
+                    .get(&StakeholderId::new(&self.leader_xpub))
+                    .unwrap();
+
+                let mut proxy_sig_message = vec!['0' as u8, '1' as u8];
+                proxy_sig_message.extend(boot_stakeholder.issuer_pk.as_ref());
+                proxy_sig_message.push(tags::SigningTag::MainBlockHeavy as u8);
+                {
+                    Serializer::new(&mut proxy_sig_message)
```

Comment:
There is no nice way in rust-cardano to create this from what I saw, since right now the blocks are created by cardano-sl haskell codebase creates the block, and rust-cardano was more for manipulating/handling them. This is just creating the proxy signature's main data/body. I will remove the braces with the next commit, I simply had some local object that I wanted to go out of scope to appease the borrow checker in there before.